java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,153 @@
+predicate isSingleHashMethodCall(MDHashMethodAccess ma) { not hasAnotherHashCall(ma) }
+
+/** Holds if `MethodAccess` ma is invoked by `MethodAccess` ma2 either directly or indirectly. */
+predicate hasParentCall(MethodAccess ma2, MethodAccess ma) {
+  ma.getCaller() = ma2.getMethod()
+  or
+  exists(MethodAccess ma3 |
+    ma.getCaller() = ma3.getMethod() and
+    hasParentCall(ma2, ma3)
+  )
+}
+
+/** Holds if `MethodAccess` is a single hashing call that is not invoked by a wrapper method. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not exists(MethodAccess ma2 | hasParentCall(ma2, ma)) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack to reduce FPs
+}
+
+/** Sink of hashing calls. */
+class HashWithoutSaltSink extends DataFlow::ExprNode {
+  HashWithoutSaltSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getAnArgument() and
+      isSink(ma)
+    )
+  }
+}
+
+/** Taint configuration tracking flow from an expression whose name suggests it holds password data to a method call that generates a hash without a salt. */
+class HashWithoutSaltConfiguration extends TaintTracking::Configuration {
+  HashWithoutSaltConfiguration() { this = "HashWithoutSaltConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source.asExpr() instanceof PasswordVarExpr }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof HashWithoutSaltSink }
+
+  /**
+   * Holds if a password is concatenated with a salt then hashed together through the call `System.arraycopy(password.getBytes(), ...)`, for example,
+   *  `System.arraycopy(password.getBytes(), 0, allBytes, 0, password.getBytes().length);`
+   *  `System.arraycopy(salt, 0, allBytes, password.getBytes().length, salt.length);`
+   *  `byte[] messageDigest = md.digest(allBytes);`
+   * Or the password is concatenated with a salt as a string.
+   */
+  override predicate isSanitizer(DataFlow::Node node) {
+    exists(MethodAccess ma |
+      ma.getMethod().getDeclaringType().hasQualifiedName("java.lang", "System") and
+      ma.getMethod().hasName("arraycopy") and
+      ma.getArgument(0) = node.asExpr()
+    ) // System.arraycopy(password.getBytes(), ...)
+    or
+    exists(AddExpr e | hasAddExpr(e, node.asExpr())) // password+salt
+    or
+    exists(ConditionalExpr ce | ce = node.asExpr()) // useSalt?password+":"+salt:password
+    or
+    exists(MethodAccess ma |
+      ma.getMethod().getDeclaringType().hasQualifiedName("java.lang", "StringBuilder") and
+      ma.getMethod().hasName("append") and
+      ma.getArgument(0) = node.asExpr() // stringBuilder.append(password).append(salt)
+    )
+    or
+    exists(MethodAccess ma |
+      ma.getQualifier().(VarAccess).getVariable().getType() instanceof Interface and
+      ma.getAnArgument() = node.asExpr() // Method access of interface type variables requires runtime determination
```

Comment:
Explain why we're sanitising all arguments to calls against an interface type?

Reply:
Hi @smowton,

Thanks for reviewing the changes. For the current implementation with `MethodAccess`, I noticed it won't be able to detect concrete implementations of interface types at runtime, e.g. the following test case:
```
	// BAD - Invoke a wrapper implementation with a salt, which is not detected with an interface type variable.
	public String getSHA256Hash7(byte[] passphrase) throws NoSuchAlgorithmException, ClassNotFoundException, IllegalAccessException, InstantiationException {
		Class c = Class.forName("SHA512");
		HASH sha512 = (HASH) (c.newInstance());
		byte[] tmp = new byte[4];
		byte[] key = new byte[32 * 2];
		for (int i = 0; i < 2; i++) {
			sha512.init();
			tmp[3] = (byte) i;
			sha512.update(passphrase, 0, passphrase.length);
			System.arraycopy(sha512.digest(), 0, key, i * 32, 32);
		}
		return Base64.getEncoder().encodeToString(key);
	}
``` 

Without the exclusion of interface types, even there is an <code>arraycopy</code> method call to mix password digest with some other bytes, the query won't detect thus will report it. However, if I change the object type from the interface type <code>HASH</code> to the class type <code>SHA512</code>, the query can detect. That is, to change
```
		Class c = Class.forName("SHA512");
		HASH sha512 = (HASH) (c.newInstance());
```  
to
```
		SHA512 sha512 = new SHA512();
```

Is there a better approach to handle this scenario? Thanks a lot.

Reply:
OK, for the time being it's probably enough to note this in a comment and leave this for when the query is promoted out of experimental.